### ISSUE TITLE
HeaderDiff doesn't recognize ignore_keywords for keyword_count

### ISSUE THREAD

**Issue by eigenbrot** [astropy/astropy#11770]
Astropy 4.2.1

Consider the following:

```
>>> h1 = fits.Header()
>>> h2 = fits.Header()
>>> h1['foo'] = 'bar'
>>> h2['foo'] = 'bar'
>>> h1['foo2'] = 'baz'
>>> diff = fits.HeaderDiff(h1, h2, ignore_keywords=['foo2'])
>>> diff.identical
False
>>> diff..diff_keyword_count
(2, 1)
```

Based on the [documentation for HeaderDiff](https://docs.astropy.org/en/stable/io/fits/api/diff.html#headerdiff) I would have expect these to be found to be identical. As shown above they are not identical because they have a different number of keywords, even though I'm trying to ignore one of them.

I did a quick check of some of the other `*Diff` methods and they seem to all suffer from the same issue.

Thanks.

**Comment by pllim** [comment]
Interestingly, this is not caught by existing tests!

https://github.com/astropy/astropy/blob/e183e398d8bdbbe5c7ccec7a793afed2ecdac4e1/astropy/io/fits/tests/test_diff.py#L239-L257

**Comment by saimn** [comment]
Indeed, even if you ignore some keyword, `diff.report()` still reports that the number of keywords is different (because `diff_keyword_count`)

```
In [5]: print(diff.report())
 Headers have different number of cards:
  a: 2
  b: 1
```

But this may be on purpose (see also https://github.com/astropy/astropy/pull/7538#discussion_r193515697), I'm not sure. Obviously the headers are not identical, and `ignore_keywords` has an effect at the card level. Should it change the answer at the header level as well ? I don't know I can see pros and cons with both solutions.

**Comment by eigenbrot** [comment]
My vote (for what it's worth) is that `ignore_keywords` should change the answer on the header level. Without that change then there is no way (at least not that I can see) of using the `diff` module to ask "are these [headers|files|etc.] the same except for these keys?". This seems like a very natural thing to want to ask. A quick example would be confirming that two headers came from the same instrument setup and are not different except for some "DATE"-like keyword.

At the very least I would say that the current behavior is unexpected based on the documentation. The `ignore_keywords` text says "...the presence of these keywords...[is] ignored". To me that sounds like it shouldn't matter whether they are there or not.

**Comment by saimn** [comment]
I understand the idea but I don't know if the current behavior was on purpose. Maybe @embray can enlighten us ? :pray:

**Comment by embray** [comment]
Taking a look...

**Comment by embray** [comment]
No recollection as to the current behavior.  The current `astropy.io.fits.diff` module was a re-implementation of an older fitsdiff.py script that was floating around years ago, I think as part of PyFITS.  It seems like the `ignore_keywords` option goes all the way back to the original re-implementation, and I don't recall whether or not it was part of the original script.

**Comment by embray** [comment]
I could go digging deeper, but my suspicion is that `ignore_keywords` just isn't taken into account in all cases.  I think it's probably used more when comparing just the values of individual keywords, but not the presence, or lack thereof, of some keyword.

I think your interpretation of how the option *should* work is valid, and patches would be welcome (though it might lead to breaking some things in some pipelines that have relied somehow on the existing behavior, but I'm open to only worrying about that if the case arises...)

**Comment by cmarmo** [comment]
Hello I have run the original snippet printing some more info about the header
```py
from astropy.io import fits

h1 = fits.Header()
h2 = fits.Header()
h1['foo'] = 'bar'
h2['foo'] = 'bar'
h1['foo2'] = 'baz'
diff = fits.HeaderDiff(h1, h2, ignore_keywords=['foo2'])
print(diff.identical)
print(diff.diff_keyword_count)
print(diff.diff_keywords)
```
the output is
```
False
(2, 1)
()
```

Apparently `diff_keyword_count` does not take into account ignored keywords while `diff_keyword` does.
This is the expected behavior from the comment in the code
https://github.com/astropy/astropy/blob/0df94ff7097961e92fd7812036a24b145bc13ca8/astropy/io/fits/diff.py#L797-L802

For instance the case reported above

> A quick example would be confirming that two headers came from the same instrument setup and are not different except for some "DATE"-like keyword.

will make use of `diff_keywords` to check if all the keywords but `DATE-something` are equal.

I'm wondering if this behavior should be modified or perhaps it is safer for back compatibility to clarify the documentation and add a test.

**Comment by cmarmo** [comment]
Referencing here #5757 and #7763 for useful discussions.
